components/tikv_util/src/config.rs

```diff
@@ -457,17 +457,84 @@ impl<'de> Deserialize<'de> for ReadableDuration {
     }
 }
 
+fn canonicalize_imp<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    fn normalize<P: AsRef<Path>>(path: P) -> PathBuf {
+        use std::path::Component;
+        let mut components = path.as_ref().components().peekable();
+        let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+            components.next();
+            PathBuf::from(c.as_os_str())
+        } else {
+            PathBuf::new()
+        };
+
+        for component in components {
+            match component {
+                Component::Prefix(..) => unreachable!(),
+                Component::RootDir => {
+                    ret.push(component.as_os_str());
+                }
+                Component::CurDir => {}
+                Component::ParentDir => {
+                    ret.pop();
```

Comment:
There is another issue without symbolic links.

If some prefix of input not exists, the following `ParentDir` cannot be parsed properly.

For example:

the current directory is: `"/folder1/folder2"`, the input is `"./not-exists/../../.."`, the output is supposed to be `"/"`, but actually we get output `""` (which means a relative path to the current directory), because it just pop the path component without checking if `ParentDir` move out of the bound.

Reply:
I think `PathBuf` already handled this case: pop on root dir is ignored.

Reply:
Emm, my statement above is kind of misleading. I mean: because the method does not canonicalize path before the loop, PahtBuf `ret` can not determine if there is a parent directory of the current directory.

A clearer example: when the current directory is `"/folder1/folder2"`, the input is `"./not-exists/../.."`, the output is supposed to be "/folder1", but actually we got output "".

Reply:
[commit#9d2fdd9](https://github.com/tikv/tikv/pull/10128/commits/9d2fdd982fb89c004a65c383f2f4c0a06e3e996a) still get the `pop` problem.

```rust
        println!(
            "current: {:?}\noutput: {:?}",
            canonicalize_imp(PathBuf::from(".")).unwrap(),
            canonicalize_imp(PathBuf::from("./not-exists/../..")).unwrap()
        );
```

output:

```
current: "/home/mrcroxx/github/tikv/tikv/components/tikv_util"
output: "/home/mrcroxx/github/tikv/tikv/components/tikv_util"
```
(which is supposed to be `"/home/mrcroxx/github/tikv/tikv/components"`)

Maybe the proper way is canonicalizing current directory first in `normalize` if `path` is relative, but it is a little bit tricky I think. Maybe we need a future discuss.

Reply:
You are right, so it seems cargo's implementation isn't so solid ;/